### PR TITLE
Fused linear projection within MLP 

### PR DESCRIPTION
**Pull request by wwwjn**

### CODE REVIEW

File: torchtitan/models/common/feed_forward.py [hunk 1]
```diff
@@ -52,3 +52,21 @@ def __init__(self, config: Config):
...
+        w13: Linear.Config
+        w2: Linear.Config
+
+    def __init__(self, config: Config):
+        super().__init__()
+        self.w13 = config.w13.build()
+        self.w2 = config.w2.build()
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        w1_out, w3_out = self.w13(x).chunk(2, dim=-1)
```
Comment (tianyu-l) [inline torchtitan/models/common/feed_forward.py:71]: `chunk` may not give great performance, IIRC. Did you benchmark this against an efficient impl, e.g. vllm?

File: torchtitan/models/llama3/parallelize.py [hunk 2]
```diff
@@ -238,11 +239,17 @@ def apply_tp(
...
                 desired_input_layouts=(Replicate(),),
             ),
-            "feed_forward.w1": colwise_parallel(),
-            "feed_forward.w2": rowwise_output_plan,
-            "feed_forward.w3": colwise_parallel(),
         }
 
+        # pyrefly: ignore [missing-attribute]
+        if isinstance(transformer_block.feed_forward, FusedFeedForward):
+            layer_plan["feed_forward.w13"] = colwise_parallel()
```
Comment (tianyu-l) [inline torchtitan/models/llama3/parallelize.py:246]: This doesn't sound right. Sharding and stacking happening in the same tensor dim causes problems, e.g. when resharding.

See the trick here https://fburl.com/code/8ou37gda

**Review by tianyu-l — CHANGES_REQUESTED** [review]

File: torchtitan/models/llama3/parallelize.py (hunk 2 above)
Comment (tianyu-l) [reply torchtitan/models/llama3/parallelize.py:246]: maybe we should wait until @fegin 's config change, before changing it away from nn.Linear, which is assumed by the current parallelize plan
